### PR TITLE
Scrollview improvements + keyboard navigation

### PR DESCRIPTION
**Pull request by GeorgeGindev**
UX behavior item - https://github.com/telerik/kendo-ux-private/issues/227, https://github.com/telerik/kendo-ux-private/issues/89
Kendo Item: https://github.com/telerik/kendo/issues/11785

### CODE REVIEW

File: packages/default/scss/scrollview/_theme.scss [hunk 1]
```diff
@@ -10,47 +10,106 @@
...
+        }
+
+
+        &:focus,
+        &.k-state-focused,
+        &.k-state-focused.k-primary {
+            border-style: solid;
+            border-width: $scrollview-pagebutton-focused-border-width;
+            border-color: $scrollview-pagebutton-focused-border-color;
+            top: -$scrollview-pagebutton-focused-border-width;
```
Comment (PreslavKozovski) [inline packages/default/scss/scrollview/_theme.scss]: everything but the `border-color` should be part of the layout.

File: packages/default/scss/scrollview/_theme.scss [hunk 2]
```diff
@@ -10,47 +10,106 @@
...
+            &.k-state-hovered {
                 opacity: $scrollview-navigation-hover-opacity;
+            }
+        }
+    }
+
+    .k-scrollview-next,
+    .k-scrollview-prev {
+        color: $scrollview-navigation-color;
+        background: $scrollview-navigation-bg;
```
Comment (PreslavKozovski) [inline packages/default/scss/scrollview/_theme.scss:37]: Use `background-color`

File: packages/default/scss/scrollview/_theme.scss [hunk 3]
```diff
@@ -10,47 +10,106 @@
...
+        text-shadow: $scrollview-navigation-icon-shadow;
+        opacity: $scrollview-navigation-default-opacity;
+        outline-width: 0;
```
Comment (PreslavKozovski) [inline packages/default/scss/scrollview/_theme.scss:40]: Move `outline-width` to `layout`

File: packages/default/scss/scrollview/_theme.scss [hunk 4]
```diff
@@ -10,47 +10,106 @@
...
+        &.k-state-focused {
+            opacity: $scrollview-navigation-hover-opacity;
+
+            span::before {
+                background-color: $scrollview-navigation-hover-span-bg;
+            }
+
+            > .k-icon::before {
+                border-style: solid;
+                border-width: $scrollview-pagebutton-focused-border-width;
```
Comment (PreslavKozovski) [inline packages/default/scss/scrollview/_theme.scss]: Move these to `layout`

File: packages/default/scss/scrollview/_layout.scss [hunk 5]
```diff
@@ -15,120 +15,132 @@
...
+        border-style: solid;
+        border-radius: 50%;
+        cursor: pointer;
+        pointer-events: all;
 
-        // remove when https://github.com/telerik/kendo-angular-scrollview/issues/130 is done
-        .k-scrollview-pageable > li.k-button::before {
-            pointer-events: initial;
-            border-radius: 0;
-        }
```
Comment (PreslavKozovski) [inline packages/default/scss/scrollview/_layout.scss:132]: No need in skipping two rows when nesting selectors in this case.

File: packages/default/scss/scrollview/_layout.scss [hunk 6]
```diff
@@ -146,19 +158,15 @@
...
-            overflow: hidden;
 
             .k-scrollview-nav {
                 margin: 0;
-                padding: ($scrollview-pager-item-spacing / 4) 0 0 0;
                 display: block;
                 white-space: nowrap;
-                overflow-x: scroll;
-                overflow-y: hidden;
                 text-align: center;
```
Comment (PreslavKozovski) [inline packages/default/scss/scrollview/_layout.scss]: This change is breaking the current logic of scrolling the pager buttons. To check it, use the following demo https://demos.telerik.com/kendo-ui/scrollview/api

File: packages/bootstrap/scss/scrollview/_theme.scss [hunk 7]
```diff
@@ -10,25 +10,91 @@
...
+
+        &:focus,
+        &.k-state-focused {
+            border-width: $scrollview-pagebutton-focused-border-width;
+            border-color: $scrollview-pagebutton-focused-border-color;
+            top: -$scrollview-pagebutton-focused-border-width;
+            &.k-primary {
+                border-color: $scrollview-pagebutton-focused-border-color;
+                border-width: $scrollview-pagebutton-focused-border-width;
+            }
```
Comment (PreslavKozovski) [inline packages/bootstrap/scss/scrollview/_theme.scss]: Move "layout" pros to the `layout` file.

File: packages/bootstrap/scss/scrollview/_theme.scss [hunk 8]
```diff
@@ -10,25 +10,91 @@
...
+        }
+    }
+
+    .k-scrollview-next,
+    .k-scrollview-prev {
+        color: $scrollview-navigation-color;
+        background: $scrollview-navigation-bg;
```
Comment (PreslavKozovski) [inline packages/bootstrap/scss/scrollview/_theme.scss:36]: use `background-color`

File: packages/bootstrap/scss/scrollview/_variables.scss [hunk 9]
```diff
@@ -8,23 +8,30 @@ $scrollview-bg: $component-bg !default;
...
-$scrollview-navigation-color: $selected-text !default;
+$scrollview-navigation-color: white !default;
 $scrollview-navigation-icon-shadow: rgba(0, 0, 0, .3) 0 0 15px !default;
 $scrollview-navigation-bg: rgba(0, 0, 0, 0) !default;
 $scrollview-navigation-default-opacity: .7 !default;
 $scrollview-navigation-hover-opacity: 1 !default;
 $scrollview-navigation-hover-span-bg: null !default;
+
+$scrollview-light-bg: rgba(255, 255, 255, .2);
+$scrollview-dark-bg: rgba(0, 0, 0, .2);
```
Comment (PreslavKozovski) [inline packages/bootstrap/scss/scrollview/_variables.scss:33]: Add `!default;` so these variables could be overridden.

File: packages/classic/scss/scrollview/_variables.scss [hunk 10]
```diff
@@ -8,23 +8,29 @@ $scrollview-bg: $component-bg !default;
...
-$scrollview-navigation-color: $selected-text !default;
+$scrollview-navigation-color: white !default;
 $scrollview-navigation-icon-shadow: rgba(0, 0, 0, .3) 0 0 15px !default;
 $scrollview-navigation-bg: rgba(0, 0, 0, 0) !default;
 $scrollview-navigation-default-opacity: .7 !default;
 $scrollview-navigation-hover-opacity: 1 !default;
 $scrollview-navigation-hover-span-bg: null !default;
+
+$scrollview-light-bg: rgba(255, 255, 255, .2);
+$scrollview-dark-bg: rgba(0, 0, 0, .2);
```
Comment (PreslavKozovski) [inline packages/classic/scss/scrollview/_variables.scss:33]: Add `!default;` so these variables could be overridden.

File: packages/default/scss/scrollview/_variables.scss [hunk 11]
```diff
@@ -8,23 +8,29 @@ $scrollview-bg: $component-bg !default;
...
-$scrollview-navigation-color: $selected-text !default;
+$scrollview-navigation-color: white !default;
 $scrollview-navigation-icon-shadow: rgba(0, 0, 0, .3) 0 0 15px !default;
 $scrollview-navigation-bg: rgba(0, 0, 0, 0) !default;
 $scrollview-navigation-default-opacity: .7 !default;
 $scrollview-navigation-hover-opacity: 1 !default;
 $scrollview-navigation-hover-span-bg: null !default;
+
+$scrollview-light-bg: rgba(255, 255, 255, .2);
+$scrollview-dark-bg: rgba(0, 0, 0, .2);
```
Comment (PreslavKozovski) [inline packages/default/scss/scrollview/_variables.scss:33]: Add `!default;` so these variables could be overridden.

File: packages/material/scss/scrollview/_variables.scss [hunk 12]
```diff
@@ -8,23 +8,29 @@ $scrollview-bg: $component-bg !default;
...
-$scrollview-navigation-color: $base-text !default;
+$scrollview-navigation-color: white !default;
 $scrollview-navigation-icon-shadow: rgba(0, 0, 0, .3) 0 0 15px !default;
 $scrollview-navigation-bg: rgba(0, 0, 0, 0) !default;
 $scrollview-navigation-default-opacity: .7 !default;
 $scrollview-navigation-hover-opacity: 1 !default;
 $scrollview-navigation-hover-span-bg: rgba($base-bg, .4) !default;
+
+$scrollview-light-bg: rgba(255, 255, 255, .2);
+$scrollview-dark-bg: rgba(0, 0, 0, .2);
```
Comment (PreslavKozovski) [inline packages/material/scss/scrollview/_variables.scss:33]: Add `!default;` so these variables could be overridden.

**Review by PreslavKozovski — CHANGES_REQUESTED** [review]
Check the below comments.

Also, in a real scenario, the pager hover does not look good. You should check why. For example: 
image

**Review by PreslavKozovski — APPROVED** [review]

**Comment by inikolova** [comment]
From my tests with the Angular ScrollView I observed the following:

- The default opacity of the pager buttons is changed;
- Hover behaviour is applied to the whole **k-scrollview-pageable** element, which changes the appearance of all buttons in the pager at the same time. Also, no hover effect for a separate page is applied;

**Review by inikolova — CHANGES_REQUESTED** [review]
From my tests with the Angular ScrollView I observed the following:

- The default opacity of the pager buttons is changed;

- Hover behaviour is applied to the whole k-scrollview-pageable element, which changes the appearance of all buttons in the pager at the same time. Also, no hover effect for a separate page is applied;

File: packages/bootstrap/scss/scrollview/_theme.scss [hunk 13]
```diff
@@ -2,34 +2,89 @@
...
+    }
+
+    kendo-scrollview.k-scrollview-wrap .k-scrollview,
+    .k-widget.k-scrollview .k-scrollview-wrap {
+        &:focus {
+            outline: none;
+
+
+        }
+    }
```
Comment (joneff) [inline packages/bootstrap/scss/scrollview/_theme.scss:41]:
```suggestion
    .k-scrollview,
    .k-scrollview-wrap {
        outline: none;
    }
```

or if that doesn't work 

```suggestion
    .k-scrollview:focus,
    .k-scrollview-wrap:focus {
        outline: none;
    }
```

File: packages/bootstrap/scss/scrollview/_theme.scss [hunk 14]
```diff
@@ -2,34 +2,89 @@
...
+    .k-scrollview-next,
+    .k-scrollview-prev {
+        color: $scrollview-navigation-color;
+        background-color: $scrollview-navigation-bg;
+        text-shadow: $scrollview-navigation-icon-shadow;
+        opacity: $scrollview-navigation-default-opacity;
+        outline-width: 0;
+
+        &:focus,
+        &.k-state-focused {
```
Comment (joneff) [inline packages/bootstrap/scss/scrollview/_theme.scss:52]: this is a new selector. prefer `k-state-focus`

File: packages/bootstrap/scss/scrollview/_theme.scss [hunk 15]
```diff
@@ -2,34 +2,89 @@
...
+            color: $scrollview-navigation-color;
+            opacity: $scrollview-navigation-hover-opacity;
+            .k-icon::before {
+                box-shadow: $scrollview-pagebutton-shadow;
+                background-color: $scrollview-navigation-hover-span-bg;
+            }
+        }
+
+        &:hover,
+        &.k-state-hovered {
```
Comment (joneff) [inline packages/bootstrap/scss/scrollview/_theme.scss:62]: this is  anew selector. prefer `k-state-hover`

File: packages/bootstrap/scss/scrollview/_theme.scss [hunk 16]
```diff
@@ -2,34 +2,89 @@
...
+            color: $scrollview-navigation-color;
+            opacity: $scrollview-navigation-hover-opacity;
+
+        }
+    }
+
+
+
+    kendo-scrollview.k-scrollview-wrap .k-scrollview-pageable > .k-button,
+    .k-scrollview .k-scrollview-nav > .k-link {
```
Comment (joneff) [inline packages/bootstrap/scss/scrollview/_theme.scss:72]:
```suggestion
    .k-scrollview-pageable > .k-button,
    .k-scrollview-nav > .k-link {
```

should be sufficient

File: packages/default/scss/scrollview/_layout.scss [hunk 17]
```diff
@@ -14,164 +14,163 @@
...
-            height: 100%;
-            cursor: default;
+        &.k-scrollview-multidot .k-scrollview-pageable > .k-button,
+        &.k-scrollview-multidot .k-scrollview-nav > .k-link {
 
-            img {
-                user-select: none;
+            &:nth-last-of-type(1),
+            &:nth-of-type(1) {
+                width: $scrollview-pagebutton-size - 6px;
```
Comment (joneff) [inline packages/default/scss/scrollview/_layout.scss:23]: do not assume the size is in pixels. use calc instead

File: packages/default/scss/scrollview/_layout.scss [hunk 18]
```diff
@@ -14,164 +14,163 @@
...
+        right: 0;
+        bottom: $scrollview-pager-offset;
+        height: $scrollview-pager-height;
+        text-align: center;
+    }
 
-        .k-scrollview-prev {
-            left: 0;
+    kendo-scrollview.k-scrollview-wrap .k-scrollview-pageable,
+    .k-scrollview .k-scrollview-nav {
```
Comment (joneff) [inline packages/default/scss/scrollview/_layout.scss:85]:
```suggestion
    .k-scrollview-pageable,
    .k-scrollview-nav {
```

Previously the selectors were shorter. use the same selectors

File: packages/default/scss/scrollview/_layout.scss [hunk 19]
```diff
@@ -14,164 +14,163 @@
...
+        &:focus,
+        &.k-state-focused {
+            outline: none;
         }
+    }
 
-        .k-scrollview-next {
-            right: 0;
+    kendo-scrollview.k-scrollview-wrap .k-scrollview-pageable > .k-button,
+    .k-scrollview .k-scrollview-nav > .k-link {
```
Comment (joneff) [inline packages/default/scss/scrollview/_layout.scss:107]:
```suggestion
    .k-scrollview-pageable > .k-button,
    .k-scrollview-nav > .k-link {
```

use shorter selectors.

File: packages/default/scss/scrollview/_layout.scss [hunk 20]
```diff
@@ -14,164 +14,163 @@
...
+        border-style: solid;
+        border-radius: 50%;
+        cursor: pointer;
+        pointer-events: all;
+        vertical-align: top;
+
+        &::before {
+            content: "";
+            width: $scrollview-pagebutton-size + $scrollview-pager-item-spacing;
+            height: $scrollview-pagebutton-size + $scrollview-pager-item-spacing;
```
Comment (joneff) [inline packages/default/scss/scrollview/_layout.scss:126]: you are assuming compatible units.

File: packages/default/scss/scrollview/_theme.scss [hunk 21]
```diff
@@ -9,48 +9,73 @@
             $scrollview-bg,
             $scrollview-border
         );
+        -webkit-tap-highlight-color: $rgba-transparent;
```
Comment (joneff) [inline packages/default/scss/scrollview/_theme.scss:12]: Why? this is not an appearance thing. it belongs to the layout.

File: packages/default/scss/scrollview/_theme.scss [hunk 22]
```diff
@@ -9,48 +9,73 @@
...
-                opacity: $scrollview-navigation-hover-opacity;
-
-                span::before {
-                    background-color: $scrollview-navigation-hover-span-bg;
-                }
+        &:focus,
+        &.k-state-focused {
+            color: $scrollview-navigation-color;
+            opacity: $scrollview-navigation-hover-opacity;
+            .k-icon::before {
```
Comment (joneff) [inline packages/default/scss/scrollview/_theme.scss:48]:
```suggestion

            .k-icon::before {
```

leave a space for legibility.

**Review by joneff — CHANGES_REQUESTED** [review]
To make things easier, consider that

* angular uses `ul.k-scrollview` and jquer uses `div.k-scrollview`
* both use `ul > li` for the navigation.

and finally, this component is getting really out of hand in terms of rendering and should be unified!

File: packages/default/scss/scrollview/_layout.scss [hunk 23]
```diff
@@ -14,164 +14,163 @@
         overflow: hidden;
         display: block;
         -webkit-touch-callout: none;
-        -webkit-tap-highlight-color: $rgba-transparent;
```
[the remark this answers is not in the document]
Comment (GeorgeGindev) [reply packages/default/scss/scrollview/_layout.scss:17]: because it is a color, I thought it belongs to theme

**Review by inikolova — CHANGES_REQUESTED** [review]
EDIT: 
In Angular in RTL mode pager buttons **::before** elements are moved, so clicking a page selects the next one:
Screenshot 2020-12-11 at 9 47 13
Screenshot 2020-12-11 at 9 47 27

**Review by inikolova — APPROVED** [review]

**Comment by inikolova** [comment]
Great job :)

**Pull request merged**